resources/tentative/complex-static-html/src/big-dom-generator/gen-css.js

```diff
@@ -1,152 +1,114 @@
 
-const getType = (depth, index) => {
-    switch (depth) {
-        case 7:
-            return "div";
-        case 6:
-            return "li";
-        case 5:
-            // prettier-ignore
-            if (!index)
-                return "div";
-            return "ul";
-        case 4:
-            // prettier-ignore
-            if (!index)
-                return "header";
-            return "main";
-        case 3:
-            return "section";
-        case 2:
-        case 1:
-            return "div";
-        case 0:
-            return ".main-ui";
+const getClassname = (element) => {
+    return element && element.classList.length > 0 ? `.${element.classList[0]}` : element.nodeName.toLowerCase();
+};
+
+const getElementType = (element) => {
+    return element.nodeName.toLowerCase();
+};
+
+const getElementAtDepth = (combinator, element, currentDepth, depth) => {
+    let currentElement = element;
+    while (currentDepth > depth) {
+        currentElement = currentElement.parentElement;
+        currentDepth--;
+    }
+    return getRandomElement(combinator, currentElement);
+};
+
+const getRandomElement = (combinator, element) => {
+    switch (combinator) {
+        case Combinator.CHILD:
+            return element;
+        case Combinator.ADJACENT_SIBLING:
+            return element.previousElementSibling;
+        case Combinator.GENERAL_SIBLING:
+            return getRandomSiblingElementBefore(element);
         default:
-            throw new Error(`Invalid depth: ${depth}`);
+            return element;
     }
 };
 
+const getRandomSiblingElementBefore = (element) => {
+    const parent = element.parentElement;
+    const children = Array.from(parent.children);
+    const currentIndex = children.indexOf(element);
+    const validChildren = children.slice(0, currentIndex);
+    return random.choice(validChildren);
+};
+
 const getNextDepth = (combinator, depth) => {
     switch (combinator) {
-        case " ":
+        case Combinator.DESCENDANT:
             return random.randRange(0, depth);
-        case " > ":
+        case Combinator.CHILD:
             return depth - 1;
-        case " + ":
-        case " ~ ":
+        case Combinator.ADJACENT_SIBLING:
+        case Combinator.GENERAL_SIBLING:
             return depth;
         default:
             throw new Error(`Invalid combinator: ${combinator}`);
     }
 };
 
-const getNextIndex = (combinator, newDepth, index) => {
-    // prettier-ignore
-    if (combinator === " + ")
-        return index - 1;
+const chooseCombinator = (depth, index) => {
+    const selectors = [Combinator.DESCENDANT, Combinator.CHILD];
     // prettier-ignore
-    if (combinator === " ~ ")
-        return random.randRange(0, index);
-    switch (newDepth) {
-        case 6:
-            return index;
-        case 5:
-        case 4:
-            return 1;
-        case 1:
-            return 4;
-        default:
-            return 0;
-    }
-};
+    if (index > 0 && depth !== 7)
```

Comment:
Can we do this with only the element? `if (element.parent.previousElementSibling)`. Do we still need `index` and `depth`?

Reply:
I tried:
```javascript
if (element.previousElementSibling)
```
This generates the exact same CSS. It will [return null if index = 0](https://developer.mozilla.org/en-US/docs/Web/API/Element/previousElementSibling), satisfying our first condition.

A slight modification to your suggestion.
```javascript
if (element.parentElement.previousElementSibling)
```
if we wanted to go with that. The reason is because `previousElementSibling` only exists on element nodes but `parent` can return any type of node.

Reply:
I tried:
```javascript
if (element.previousElementSibling)
```
This generates the exact same CSS. It will [return null if index = 0](https://developer.mozilla.org/en-US/docs/Web/API/Element/previousElementSibling), satisfying our first condition.

A slight modification to your suggestion. if we wanted to go with that.
```javascript
if (element.parentElement.previousElementSibling)
``` 
This code generates different css.